File: src/main/java/fi/quanfoxes/parser/patterns/InstructionPattern.java
package fi.quanfoxes.parser.patterns;

import java.util.List;

import fi.quanfoxes.Keyword;
import fi.quanfoxes.Keywords;
import fi.quanfoxes.lexer.KeywordToken;
import fi.quanfoxes.lexer.Token;
import fi.quanfoxes.lexer.TokenType;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Pattern;
import fi.quanfoxes.parser.nodes.InstructionNode;

public class InstructionPattern extends Pattern {
    public static final int PRIORITY = 1;

    private static final int INSTRUCTION = 0;

    public InstructionPattern() {
        // Pattern:
        // continue, stop
        super(TokenType.KEYWORD);
    }

    @Override
    public int priority(List<Token> tokens) {
        return PRIORITY;
    }

    @Override
    public boolean passes(List<Token> tokens) {
        KeywordToken keyword = (KeywordToken)tokens.get(INSTRUCTION);
        return keyword.getKeyword() == Keywords.CONTINUE || keyword.getKeyword() == Keywords.STOP;
    }

    private Keyword getInstruction(List<Token> tokens) {
        KeywordToken instruction = (KeywordToken)tokens.get(INSTRUCTION);
        return instruction.getKeyword();
    }

	@Override
	public Node build(Context context, List<Token> tokens) throws Exception {
		return new InstructionNode(getInstruction(tokens));
	}
}

File: Zigzag/Parser/Patterns/InstructionPattern.cs
using System.Collections.Generic;

public class InstructionPattern : Pattern
{
	public const int PRIORITY = 1;

	private const int INSTRUCTION = 0;

	// Pattern:
	// continue, stop
	public InstructionPattern() : base(TokenType.KEYWORD) { }

	public override int GetPriority(List<Token> tokens)
	{
		return PRIORITY;
	}

	public override bool Passes(List<Token> tokens)
	{
		KeywordToken keyword = (KeywordToken)tokens[INSTRUCTION];
		return keyword.Keyword == Keywords.CONTINUE || keyword.Keyword == Keywords.STOP;
	}

	private Keyword GetInstruction(List<Token> tokens)
	{
		KeywordToken instruction = (KeywordToken)tokens[INSTRUCTION];
		return instruction.Keyword;
	}

	public override Node Build(Context context, List<Token> tokens)
	{
		return new InstructionNode(GetInstruction(tokens));
	}
}